subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/NestedBeanAnnotationHandler.java

```diff
@@ -34,7 +41,35 @@ public boolean isPropertyRelevant() {
         return true;
     }
 
+    @Override
+    @SuppressWarnings("unchecked")
+    public void validatePropertyMetadata(PropertyMetadata propertyMetadata, TypeValidationContext validationContext) {
+        if (Map.class.isAssignableFrom(propertyMetadata.getDeclaredType().getRawType())) {
+            Class<?> keyType = extractNestedType((TypeToken<Map<?, ?>>) propertyMetadata.getDeclaredType(), Map.class, 0);
+            validateKeyType(propertyMetadata, validationContext, keyType);
+        }
+    }
+
     @Override
     public void visitPropertyValue(String propertyName, PropertyValue value, PropertyMetadata propertyMetadata, PropertyVisitor visitor) {
     }
+
+    private static <T> Class<?> extractNestedType(TypeToken<T> beanType, Class<? super T> parameterizedSuperClass, int typeParameterIndex) {
+        ParameterizedType type = (ParameterizedType) beanType.getSupertype(parameterizedSuperClass).getType();
+        return TypeToken.of(type.getActualTypeArguments()[typeParameterIndex]).getRawType();
+    }
```

Comment:
:x: Let's extract this method to some utility type; we copied this over from `AbstractTypeMetadataWalker`. One option is to create a new utility type in this package, but I think it would make sense to add it in `JavaReflectionUtil` in `:base-services`.